index.js

```diff
@@ -2127,6 +2077,62 @@
                    foldable);
   }
 
+  //# takeWhile :: (Applicative f, Foldable f, Monoid (f a)) => (a -> Boolean, f a) -> f a
+  //.
+  //. Discards the first element which does not satisfy the predicate, and all
+  //. subsequent elements.
+  //.
+  //. This function is derived from [`concat`](#concat), [`empty`](#empty),
+  //. [`of`](#of), and [`reduce`](#reduce).
+  //.
+  //. See also [`dropWhile`](#dropWhile).
+  //.
+  //. ```javascript
+  //. > takeWhile (s => /x/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xy', 'xz', 'yx']
+  //.
+  //. > takeWhile (s => /y/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xy']
+  //.
+  //. > takeWhile (s => /z/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. []
+  //. ```
+  function takeWhile(pred, foldable) {
+    var F = foldable.constructor;
+    var take = true;
+    return reduce (function(xs, x) {
+      return ((take = take && pred (x))) ? concat (xs, of (F, x)) : xs;
```

Comment:
I'm still a little bit uneasy about the `take = take && ...` side effect implicitly going on in here, although I can't formalize precisely why. I'll think about it a little bit.

Reply:
It makes me uneasy too, but passing around `{ take :: Boolean, xs :: f a }` values is inelegant. Would you feel equally uneasy were `take` stored in the accumulator?

Reply:
@davidchambers No, I think putting it into the zero for the fold and always pulling it off the `xs` for the fold would make me happy, because it's basically equivalent to a traversal wrt `State Boolean`.

Reply:
I guess this is extremely pathological, but is the following implementation of `reduce` for arrays illegal wrt `Foldable`'s laws?

```
// :: (b -> a -> b) -> b -> [a] -> b
const reduce = f => z => xs => {
  const _ = xs.reduceRight((p, c) => f(p)(c), z)
  return xs.reduce((p, c) => f(p)(c), z)
}
```

Reply:
### Option :tada:: Variable reassignment

```javascript
function takeWhile(pred, foldable) {
  var F = foldable.constructor;
  var take = true;
  return reduce (function(xs, x) {
    return ((take = take && pred (x))) ? concat (xs, of (F, x)) : xs;
  }, empty (F), foldable);
}
```

### Option :heart:: Pure accumulator

```javascript
function takeWhile(pred, foldable) {
  var F = foldable.constructor;
  return (reduce (function(acc, x) {
    var take = acc.take && pred (x);
    return {take: take, xs: take ? concat (acc.xs, of (F, x)) : acc.xs};
  }, {take: true, xs: empty (F)}, foldable)).xs;
}
```

### Option :rocket:: Impure accumulator

```javascript
function takeWhile(pred, foldable) {
  var F = foldable.constructor;
  return (reduce (function(r, x) {
    r.xs = (r.take = r.take && pred (x)) ? concat (r.xs, of (F, x)) : r.xs;
    return r;
  }, {take: true, xs: empty (F)}, foldable)).xs;
}
```

Which option do you favour, Asad? Which option do others favour?

Reply:
I prefer the variable reassignment for low-level utilities. There's not object creation overhead as there is for option :heart: and no property access overhead as there is for option :rocket:.

I don't have benchmarks to support my preference though.

I would never do this in application code unless it was on a hot path though 😄

Reply:
Thanks for your input, Gabe. :)

Asad makes a good point, which is that it's not necessarily safe to provide an impure function to `reduce`. Consider this pathological case:

```javascript
Array.prototype['fantasy-land/reduce'] = function(f, x) {
  return this.reduce (f, x), this.reduce (f, x);
};
```

```javascript
var take = true;
Z.reduce ((xs, x) => (take = take && x <= 3) ? Z.concat (xs, [x]) : xs,
          [],
          [1, 2, 3, 4, 5]);
```

This will evaluate to `[]` rather than `[1, 2, 3]` because evaluating the first occurrence of `this.reduce (f, x)` will set `take` to `false`.

Is this `Array#fantasy-land/reduce` implementation sensible? No. It is lawful, though, so it should interoperate correctly with `Z` functions.

Reply:
If we do want to do the efficient mutatey thing I'd be happier if it was an explicit traversal wrt `State` (not necessarily the pure `State`), because that wouldn't have these weird edge cases.

Reply:
I don't know what you mean exactly, Asad. Are you saying you like :rocket: more than :tada:?

Reply:
@davidchambers No, sorry, what I mean is that while :heart: is the option I like best of the ones you presented, I realize that it's inefficient.

If we want to use mutationto make this more efficient (while still preserving correctness), I would prefer if we captured this explicitly by using `traverse` wrt a state mutation effect.

This doesn't implicitly interleave side effects with the arbitrary typeclass implementations, the pathological scenario we were discussing earlier would still work correctly, and you still avoid any efficiency problems with the :heart: approach (whatever they may be, I haven't actually measured either).

Reply:
Are you suggesting we use `traverse` rather than `reduce`? Could you suggest an implementation?

Reply:
@davidchambers I tried implementing it but the best I could do was:

```
// :: Witherable f -> (a -> Boolean) -> f a -> f a
const takeWhile = W => test => {
  // :: a -> State Boolean (Maybe a)
  const action = a => stop =>
    stop || !test(a) ? [Nothing, false] : [Just(a), true];

  return Fn.pipe([
    W.wither(State)(action),
    State.eval
  ]);
};
```

where `type Witherable t = (Traversable t) & { wither: Applicative f -> (a -> f (Maybe b)) -> t a -> f (t b) }`. I feel like there might be a way to do it using only `Traversable` but I couldn't quite figure it out.

While writing this, I looked back at the implementation we have now and started wondering what would happen if I tried to `takeWhile` a `ZipList` or an `Observable` (where the `of` produces an infinite "repetition" of the value). We should probably try out the implementation on a few different sorts of things, or if we only plan to support arrays we should just specialize it to arrays.